📝 Add docs for contributing to the dataservice (#683)
kids-first/kf-api-dataservice: CONTRIBUTING.md

# 👥 Contributing to the Kids First Dataservice API

## Adding fields

> The dataservice is architected so that fields may be added to as needs arise.

Reviewer: ```suggestion
Over time new data comes into the Kids First ecosystem, required the Dataservice to adapt. In these cases, developers need to add new columns or change columns in the Dataservice DB and API.
```

### Where to Add Fields

> For a new field in an existing endpoint in the dataservice, those new fields are

Reviewer: ```suggestion
In order to add a new field to the Dataservice, the developer needs to know which table to add the field to. All entity types in the Dataservice have a standard folder structure like this:

dataservice/api/genomic_file
|-- README.md
|-- __init__.py
|-- models.py
|-- resources.py
`-- schemas.py

`models.py` is where the table schema is defined, using the SQLAlchemy ORM (object relational mapper). 

There are 2 things you need to modify in this file.

1. Please add a parameter and type description for the new field in the class's docstring
2. Add the field using SQLAlchemy's Column object. Take a look at other fields as an example. 

Column data types can be found in [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/20/core/type_basics.html#generic-camelcase-types),
but the most common data types used in the dataservice are `Text`, `Boolean`,
`Integer`, `Float`, and `DateTime`.
```

                "invalid": [
                    { "paired_end": -1 },
                    ...
                    { "new_field_name": "some invalid value" }
                ]
            }
            ...
        }
    }
    ```

3. `tests/create/py` **biospecimens only**

    For biospecimens only, add the new field with a valid value to the body
    object in the function `make_biospecimen`.

4. `tests/[endpoint_name]/test_[endpoint_name]_models.py`

    Testing has changed over time so the format of this file depends on the how
    testing was performed when the testing script was written. In this file,
    look for a place where a dictionary object is created with key/value pairs
    for all of the endpoint's fields and add a new, relevant key/value pair for
    the new field.

5. `tests/[endpoint_name]/test_[endpoint_name]_resources.py`

    Testing has changed over time so the format of this file depends on the how
    testing was performed when the testing script was written. In this file,
    look for a place where a dictionary object is created with key/value pairs
    for all of the endpoint's fields and add a new, relevant key/value pair for
    the new field.

> ### Migrating

Reviewer: ```suggestion
### Database Migration
```

Reviewer: This section should come before the test section otherwise the tests will fail

> After adding the new field and adding testing, the last thing to do is add a

Reviewer: ```suggestion
Now that you've added the new column to the entity's model class, you will need to generate an actually database migration script which when run will actually modify the database and it's schema to add the new column. 

To understand how the SQLAlchemy model object and database table are related you can read more about the SQLAlchemy ORM. 
```

migration script so that any migrating that needs to be done for fields can be
completed. This step must be completed for every newly added field.

From the root of the repository, run:

To create the migration:

```sh
flask db migrate -m "name of migration"
```

Then, to apply the migration:

```sh
flask db upgrade
```

> Last, add the semantic version number to the docstring of the `migration.py`

Reviewer: ```suggestion
Last, add the semantic version number to the docstring of the generated migration file
```

file.

## Pull Requests

After making changes, make sure to format your commit messages using the
guidance provided by the
[KF Developer Handbook](https://kids-first.github.io/kf-developer-handbook/guides/commit_messages.html).
Then, push your changes and make a pull request (again following those
guidelines). Make sure that all of the linting and tests pass and then request
a review on the PR by @dataservice-reviewers. Once approved, merge the PR and
delete the feature branch.

## Releases

> After merging a PR to the dataservice, Jenkins picks up the change a rebuilds

Reviewer: ```suggestion
After merging a PR into the master branch of the dataservice Github repository, Jenkins picks up the change and rebuilds the Dataservice docker image. This image is then deployed into the QA environment.
```

# 👥 Contributing to the Kids First Dataservice API

## Adding fields

> Over time new data comes into the Kids First ecosystem, required the

Reviewer: ```suggestion
Over time, new data comes into the Kids First ecosystem, requiring the
```

Dataservice to adapt. In these cases, developers need to add new columns or
change columns in the Dataservice DB and API.

### Where to Add Fields

In order to add a new field to the Dataservice, the developer needs to know
which table to add the field to. All entity types in the Dataservice have a
standard folder structure like this:

> dataservice/api/genomic_file

Reviewer: This should be enclosed in a code block otherwise it doesn't format properly. So like 
```
fill tree here
```

|-- README.md
|-- __init__.py
|-- models.py
|-- resources.py
`-- schemas.py

`models.py` is where the table schema is defined, using the SQLAlchemy ORM
(object relational mapper). To understand how the SQLAlchemy model object and
database table are related you can read more about the SQLAlchemy ORM.

There are 2 things you need to modify in this file.

1. Please add a parameter and type description for the new field in the
   class's docstring
2. Add the field using SQLAlchemy's Column object. Take a look at other fields
   as an example.

Column data types can be found in [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/20/core/type_basics.html#generic-camelcase-types),
but the most common data types used in the dataservice are `Text`, `Boolean`,
`Integer`, `Float`, and `DateTime`.

> As an example, take adding a new field, `my_new_field` to the `genomic_file`

Reviewer: ```suggestion
### Example

As an example, take adding a new field, `my_new_field` to the `genomic_file`
```

                "invalid": [
                    { "paired_end": -1 },
                    ...
                    { "new_field_name": "some invalid value" }
                ]
            }
            ...
        }
    }
    ```

3. `tests/create/py` __biospecimens only__

    For biospecimens only, add the new field with a valid value to the body
    object in the function `make_biospecimen`.

4. `tests/[endpoint_name]/test_[endpoint_name]_models.py`

    Testing has changed over time so the format of this file depends on the how
    testing was performed when the testing script was written. In this file,
    look for a place where a dictionary object is created with key/value pairs
    for all of the endpoint's fields and add a new, relevant key/value pair for
    the new field.

5. `tests/[endpoint_name]/test_[endpoint_name]_resources.py`

    Testing has changed over time so the format of this file depends on the how
    testing was performed when the testing script was written. In this file,
    look for a place where a dictionary object is created with key/value pairs
    for all of the endpoint's fields and add a new, relevant key/value pair for
    the new field.

> ### Database Migration

Reviewer: This section should be above the testing section, otherwise tests will fail

which table to add the field to. All entity types in the Dataservice have a
standard folder structure like this:

```rtf
dataservice/api/genomic_file
|-- README.md
|-- __init__.py
|-- models.py
|-- resources.py
`-- schemas.py
```

`models.py` is where the table schema is defined, using the SQLAlchemy ORM
(object relational mapper). To understand how the SQLAlchemy model object and
database table are related you can read more about the SQLAlchemy ORM.

There are 2 things you need to modify in this file.

1. Please add a parameter and type description for the new field in the
   class's docstring
2. Add the field using SQLAlchemy's Column object. Take a look at other fields
   as an example.

Column data types can be found in [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/20/core/type_basics.html#generic-camelcase-types),
but the most common data types used in the dataservice are `Text`, `Boolean`,
`Integer`, `Float`, and `DateTime`.

### Example

As an example, take adding a new field, `my_new_field` to the `genomic_file`
endpoint. This is a free text field.

In the file `dataservice/api/genomic_file/models.py`, add the docstring for
the new field:

```python

>     :param new_field_name: Description of the new field.

Reviewer: ```suggestion
    :param new_field_name: Description of the new field.
    :type new_field_name: <type of new_field_name>
```
